commons-mongo/src/main/scala/com/avsystem/commons/mongo/typed/TypedMongoCollection.scala

```diff
@@ -0,0 +1,292 @@
+
+  def deleteOne(
+    filter: MongoDocumentFilter[E],
+    setupOptions: DeleteOptions => DeleteOptions = identity
+  ): Task[DeleteResult] =
+    single(nativeCollection.deleteOne(filter.toBson, setupOptions(new DeleteOptions)))
+
+  def deleteMany(
+    filter: MongoDocumentFilter[E],
+    setupOptions: DeleteOptions => DeleteOptions = identity
+  ): Task[DeleteResult] =
+    single(nativeCollection.deleteMany(filter.toBson, setupOptions(new DeleteOptions)))
+
+  def updateOne(
+    filter: MongoDocumentFilter[E],
+    update: MongoDocumentUpdate[E],
+    upsert: Boolean = false, // extracted as separate param because it's very commonly used
+    setupOptions: UpdateOptions => UpdateOptions = identity
+  ): Task[UpdateResult] = {
+    val options = setupOptions(new UpdateOptions).upsert(upsert)
+    val (updateBson, arrayFilters) = update.toBsonAndArrayFilters
+    if (!arrayFilters.isEmpty) {
+      options.arrayFilters(arrayFilters)
+    }
+    single(nativeCollection.updateOne(filter.toBson, updateBson, options))
+  }
+
+  def updateMany(
+    filter: MongoDocumentFilter[E],
+    update: MongoDocumentUpdate[E],
+    upsert: Boolean = false, // extracted as separate param because it's very commonly used
+    setupOptions: UpdateOptions => UpdateOptions = identity
+  ): Task[UpdateResult] = {
+    val options = setupOptions(new UpdateOptions).upsert(upsert)
+    val (updateBson, arrayFilters) = update.toBsonAndArrayFilters
+    if (!arrayFilters.isEmpty) {
+      options.arrayFilters(arrayFilters)
+    }
+    single(nativeCollection.updateMany(filter.toBson, updateBson, options))
+  }
+
+  def replaceOne(
+    filter: MongoDocumentFilter[E],
+    replacement: E,
+    upsert: Boolean = false, // extracted as separate param because it's very commonly used
+    setupOptions: ReplaceOptions => ReplaceOptions = identity
+  ): Task[UpdateResult] = {
+    val options = setupOptions(new ReplaceOptions).upsert(upsert)
+    single(nativeCollection.replaceOne(filter.toBson, replacement, options))
+  }
+
+  def bulkWrite(
+    writes: Seq[MongoWrite[E]],
+    setupOptions: BulkWriteOptions => BulkWriteOptions = identity
+  ): Task[BulkWriteResult] = {
+    val requests = writes.iterator.map(_.toWriteModel).to[JList]
+    single(nativeCollection.bulkWrite(requests, setupOptions(new BulkWriteOptions)))
+  }
+
+  def createIndex(
+    index: MongoIndex[E],
+    setupOptions: IndexOptions => IndexOptions = identity
+  ): Task[String] =
+    single(nativeCollection.createIndex(index.toBson, setupOptions(new IndexOptions)))
+
+  def createIndexes(
+    indexes: Seq[(MongoIndex[E], IndexOptions)],
+    setupOptions: CreateIndexOptions => CreateIndexOptions = identity
+  ): Task[String] = {
+    val indexModels = indexes.iterator.map {
+      case (index, options) => new IndexModel(index.toBson, options)
+    }.to[JList]
+    single(nativeCollection.createIndexes(indexModels, setupOptions(new CreateIndexOptions)))
+  }
+}
```

Comment:
Would feel so good to have a nicer `.aggregate` API (or at least validate it fits with your model in the future). It's easily the ugliest, least safe Mongo code we have in our codebase.

Reply:
Yeah, I haven't looked into it yet but I'm not sure if we can express pipelines in any better-typed way.